Recipe for reading `uint` from contract (#48)
scaffold-eth/se2-docs: docs/recipes/ReadUintFromContract.md

---
sidebar_position: 3
title: Read a uint from a contract
description: Learn how to read from contract functions which accepts arguments / no arguments and display them on UI.
---

> 

Author: Ohh yes, missed it tysm 🙌, fixed it [831f768](https://github.com/scaffold-eth/se2-docs/pull/48/commits/831f768fa7dc24ca258d7c2aaeb14c72e8643337)

> This recipe demonstrates how to read data from contract functions that accept some arguments (parameters) or no arguments at all and displays it on the UI.

Reviewer: Sometimes when I read this sentence it looks strange, maybe because it's too large? (Other times I read it and seems OK heh).

What about: "This recipe demonstrates how to read data from contract functions and display it on the UI. We'll showcase and example that accept some arguments (parameters), and another with or no arguments at all"

@edakturk14 could you also take a look with your writer's eyes? 🙃

Author: > It's much more clear + readable :D

💯💯,  Tysm Pablo 🙌

        <div className="card-actions items-center flex-col gap-1 text-lg">
          <h2 className="font-bold m-0">Total Greetings count:</h2>
          {isTotalCounterLoading ? (
            <span className="loading loading-spinner"></span>
          ) : (
            <p className="m-0">{totalCounter ? totalCounter.toString() : 0}</p>
          )}
          <h2 className="font-bold m-0">Your Greetings count:</h2>
          {isConnectedAddressCounterLoading ? (
            <span className="loading loading-spinner"></span>
          ) : (
            <p className="m-0">{connectedAddressCounter ? connectedAddressCounter.toString() : 0}</p>
          )}
        </div>
      </div>
    </div>
  );
};
```

</details>

## Implementation guide

### Step 1: Create a new Component

Begin by creating a new component in the "components" folder of your application.

```tsx title="components/GreetingsCount.tsx"
export const GreetingsCount = () => {
  return (
    <div>
      <h2 className="font-bold m-0">Total Greetings count:</h2>
      <h2 className="font-bold m-0">Your Greetings count:</h2>
    </div>
  );
};
```

### Step 2: Retrieve total greetings count

> Initialize the [useScaffoldContractRead](/hooks/scaffold-eth#usescaffoldcontractread) hook to read from the contract. This hook provides the `data` which contains the return value of the function.

Author: Oops, thanks 🙌

```tsx title="components/GreetingsCount.tsx"
//highlight-start
import { useScaffoldContractRead } from "~~/hooks/scaffold-eth";
// highlight-end

export const GreetingsCount = () => {
  // highlight-start
  const { data: totalCounter } = useScaffoldContractRead({
    contractName: "YourContract",
    functionName: "totalCounter",
    watch: true,
  });
  // highlight-end

  return (
    <div>
      <h2 className="font-bold m-0">Total Greetings count:</h2>
      //highlight-start
      <p>{totalCounter ? totalCounter.toString() : 0}</p>
      //highlight-end
      <h2 className="font-bold m-0">Your Greetings count:</h2>
    </div>
  );
};
```

`const {data: totalCounter} = useScaffoldContractRead({...})` here we are using [destructuring asssignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) to assign `data` to new name `totalCounter`.

> `totalCounter` in contract returns an `uint` value, which is represented as [`BigInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) in javascript and can be converted to readable string using `.toString()`.

Reviewer 2: ```suggestion
In the contract, `totalCounter` returns an `uint` value, which is represented as a [`BigInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) in javascript and can be converted to a readable string using `.toString()`.
```

### Step 3: Retrieve connected address greetings count

> We can access connected address using [useAccount](https://wagmi.sh/react/hooks/useAccount) hook and pass it to `args` key in `useScaffoldContractRead` hooks configuration which will be used as an argument to the read contract function.

Reviewer 2: ```suggestion
We can get the connected address using the [useAccount](https://wagmi.sh/react/hooks/useAccount) hook and pass it to `args` key in the `useScaffoldContractRead` hooks configuration. The address will then be used as an argument for reading data from the contract function.
```

^maybe changed the wording too much here, and made it perhaps longer. We can also say "This will be used as an argument to read the contract function." The inital wording felt a bit confusing to me.

Author: Yup fixed at [77690fe](https://github.com/scaffold-eth/se2-docs/pull/48/commits/77690fe10d3a712ad431b6c1db6a1fd28ec640fc), thanks 🙌

```tsx title="components/GreetingsCount.tsx"
import { useScaffoldContractRead } from "~~/hooks/scaffold-eth";
//highlight-start
import { useAccount } from "wagmi";
//highlight-end

export const GreetingsCount = () => {
  //highlight-start
  const { address: connectedAddress } = useAccount();
  //highlight-end

  const { data: totalCounter } = useScaffoldContractRead({
    contractName: "YourContract",
    functionName: "totalCounter",
    watch: true,
  });

  //highlight-start
  const { data: connectedAddressCounter } = useScaffoldContractRead({
    contractName: "YourContract",
    functionName: "userGreetingCounter",
    args: [connectedAddress], // passing args to function
    watch: true,
  });
  //highlight-end

  return (
    <div>
      <h2>Total Greetings count:</h2>
      <p>{totalCounter ? totalCounter.toString() : 0}</p>
      <h2>Your Greetings count:</h2>
      //highlight-start
      <p>{connectedAddressCounter ? connectedAddressCounter.toString() : 0}</p>
      //highlight-end
    </div>
  );
};
```

### Step 4: Bonus adding loading state

> We can use `isLoading` returned from [`useScaffoldContractRead`](/hooks/scaffold-eth#usescaffoldcontractread) which is set to `true` while fetching the data from contract.

Reviewer 2: ```suggestion
We can use `isLoading` returned from the [`useScaffoldContractRead`](/hooks/scaffold-eth#usescaffoldcontractread) hook. This variable is set to `true` while fetching data from the contract.
```
